### PR TITLE
Enhancement demultiplex test

### PR DESCRIPTION
**Pull request by mstapylton**
# Pull Request Template for MMEDS

## What has changed
mmeds/tests/unit/test_demultiplex.py
    main script for testing demultiplexing, needs to be expanded to other demultiplexing options but that should be easy now that the framework is in place.

mmeds/util.py
    some helper functions here.
    validate_demultiplex should arguably be called during Qiime2 analysis, can help us catch problems in analyses. 

## Checklist of pre-requisites
-   [x] Does the code run?  
-   [x] Does the code follow the repository style?  
-   [x] Is the code tested?  

## How to use the feature
Test barcodes, reads can be found here, they don't fit on github. Need to find a better solution:
xxx_{file} 
`/sc/arion/projects/MMEDS/mmeds_server_data/adamcantor22_All_Comer_Bladder_Cancer_Patients_Final_2_0`

Here's where  mmed's module files currently are:
`/sc/arion/projects/MMEDS/.modules`

The ones for qiime1 and pheniqs need to be updated from here in this branch:
`/sc/arion/projects/MMEDS/mmeds-stable/modules`

I successfully ran validate_demultiplex in a qiime1 environment locally, by copying the files from minerva and editing the script headers in the bin of the conda env folder:
Screenshot from 2021-10-26 09-20-40

Here's the location of the qiime1 environment
`/hpc/packages/minerva-centos7/anaconda2/2019.03/envs/qiime1.9.1`

## How the design has changed
       def test_strip_error_barcodes(self): is from the original test case Adam put together, I also built the pheniqs test case off of his

### CODE REVIEW

File: mmeds/tests/unit/test_demultiplex.py
Comment (adamcantor22) [inline mmeds/tests/unit/test_demultiplex.py:56]: change `cls` to `self` in this method

File: mmeds/tests/unit/test_demultiplex.py
Comment (adamcantor22) [inline mmeds/tests/unit/test_demultiplex.py:115]: this is formatted a little weird, just make the newlines look a little nicer. but also, what would it mean if this error was caught? should the test fail if it does?

File: mmeds/util.py
Comment (adamcantor22) [inline mmeds/util.py:949]: oof. it's on my to do list to do a whole rewrite of this like i did with `strip_error_barcodes`

File: mmeds/util.py
Comment (adamcantor22) [inline mmeds/util.py:1273]:
```suggestion
    if on_chimera:
        validate_demux_file = [
                             'ml python/2.7.9-UCS4;',
                             'validate_demultiplexed_fasta.py -b -a',
                             '-i {}'.format(test_file),
                             '-o {}'.format(log_dir),
                             '-m {};'.format(map_file),
                             'rm {};'.format(test_file)
                 ]
```
do something like this instead, it's more consistent with the codebase and those backslashes look clunky. then when you need to run it do you also need `validate_demux_file.join(' ')`? I would assume you do but it doesn't look like you needed that with the other commands you wrote as lists

also use their full option names instead of the single letter if possible

File: mmeds/util.py
Comment (adamcantor22) [inline mmeds/util.py:1292]: i've warmed up to python a lot in the past few months, but it's gross that the return type isn't defined and you can just do this lol

File: scripts/validate_demux_files.py
Comment (adamcantor22) [inline scripts/validate_demux_files.py:18]: is this really the only one that's required? it seems like all of them are, currently

File: scripts/validate_demux_files.py
Comment (adamcantor22) [inline scripts/validate_demux_files.py:50]: is this necessary? is the `.gz` not getting removed when it's gunzipped?

File: scripts/validate_demux_files.py
Comment (adamcantor22) [inline scripts/validate_demux_files.py:59]: if it's gzipped do you not need to run gunzip on it?

**Review by adamcantor22 — CHANGES_REQUESTED** [review]
Some things to modify, feel free to answer the questions as responses to the comments

File: scripts/validate_demux_files.py
Comment (mstapylton) [reply scripts/validate_demux_files.py:59]: I do, is_gzip is passed to validate. The validator is run on blah.fastq though not blah.fastq.gz so I make the change here

File: scripts/validate_demux_files.py
Comment (mstapylton) [reply scripts/validate_demux_files.py:50]: it is on the file system, but that won't automatically happen for the file name in the code

File: scripts/validate_demux_files.py
Comment (mstapylton) [reply scripts/validate_demux_files.py:18]: yeah i should fix, thanks

File: mmeds/util.py
Comment (mstapylton) [reply mmeds/util.py:1292]: ;) I'd rather do multiple returns but that isn't pythonic

File: mmeds/config.py
Comment (mstapylton) [reply mmeds/config.py:50]: sure for the comments, the TODO is not in the purview of this PR

File: mmeds/config.py
Comment (mstapylton) [reply mmeds/config.py:582]: config.py doesn't try to connect to the sql or mongo dbs
mmeds imports hard coded values/paths from config.py but config.py also connects to the database, which causes issues for any scripts trying to import those values/paths without launching the full web app.

File: mmeds/server.py
Comment (mstapylton) [reply mmeds/server.py:443]: oh yeah, forgot about this. Do want the try catch to log the exception though, otherwise the validator doesn't actually report what column is causing the validation error

File: mmeds/tests/unit/test_demultiplex.py
Comment (mstapylton) [reply mmeds/tests/unit/test_demultiplex.py:115]: hmm I think the logic around doing subprocesses in try-excepts is that otherwise it's hard to get info if they do fail. Don't think the test should fail if this fails, also what newlines?

File: mmeds/util.py
Comment (mstapylton) [reply mmeds/util.py:949]: :)

File: mmeds/util.py
Comment (mstapylton) [reply mmeds/util.py:1273]: booo no we like fstrings over .format. .format is the old way! backslashes in this case are needed to split across multiple lines.
I wrote the commands as lists because that's how I found David did it, sometimes that caused issues though and then I did it as a single string rather than a list of strings.
I'll find the full option names and update

Comment (adamcantor22) [reply mmeds/util.py:1273]: lol, it's fine to use fstrings then, but still use a list like this instead of one big fstring

Comment (mstapylton) [reply mmeds/util.py:1273]: nah need one big fstring :)
having it as a list does break the code, I spent toooo long trying to make sure this worked

File: mmeds/server.py
Comment (mstapylton) [reply mmeds/server.py:443]: the exception does get handled in the next block, but the try except makes sure it gets logged. Adding a todo because this is sub-optimal

File: mmeds/tests/unit/test_demultiplex.py
Comment (adamcantor22) [inline mmeds/tests/unit/test_demultiplex.py:112]: still need to change `cls` to `self`

File: mmeds/util.py
Comment (adamcantor22) [inline mmeds/util.py:949]: Not gonna force you to write more tests here, but the `util.py` changes are what's causing the codecov to fail. Are there any tests for these new functions?

**Review by adamcantor22 — APPROVED** [review]
Approved if you don't have time to do either of these, but take a look. It's just changing `cls` to `self` in that one method, and also pointing out where the codecov is failing

**Review by adamcantor22 — CHANGES_REQUESTED** [review]
Actually, before you can merge, see if you can fix the Code Factor problem. It says using `shell=True` is a security issue, we don't want that to get merged in if at all possible

**Review by adamcantor22 — APPROVED** [review]

**Pull request merged**
